### PR TITLE
Fix clang 5 osx

### PR DESCRIPTION
**Pull request by rwols**
The plugin didn't detect clang 5 on macOS here. This PR should fix that. It should also distinguish between clang and AppleClang now.

### CODE REVIEW

**Comment by niosus** [comment]
Hey @rwols thanks for this contribution. However, I wanted to avoid using libclang bindings for clang 5.0 until clang 5.0 becomes official. Unfortunately, I don't have the time to maintain the version that is subject to change and I suggest currently sticking to clang 4.0 on all systems. Also, I am running slightly modified `cindexXX.py` files. I have changed the way they load the library as the default version was not working on all systems I wanted the plugin to run on. See [`cindex40.py`](https://github.com/niosus/EasyClangComplete/blob/master/plugin/clang/cindex40.py#L3866) as an example. 

I believe your original issue comes from the fact that I was lazy when I was implementing the logic for finding out if we are running apple version of clang, thus the TODO that you might have seen in the code. I would suggest fixing that and then the old bindings for libclang 4.0 should still work as expected with the new clang.

**Comment by rwols** [comment]
Hi @niosus, thanks for the reply.

> I believe your original issue comes from the fact that I was lazy when I was implementing the logic for finding out if we are running apple version of clang, thus the TODO that you might have seen in the code. I would suggest fixing that and then the old bindings for libclang 4.0 should still work as expected with the new clang.

I'll try this out, stay tuned.

**Comment by rwols** [comment]
You were right, cindex40 works with clang 5 too, at least judging by some simple tests.

File: plugin/tools.py [hunk 1]
```diff
@@ -684,8 +684,8 @@ def run_command(command, shell=True):
             log.debug("clang process output: \n%s", output_text)
         return output_text
 
-    @staticmethod
-    def get_clang_version_str(clang_binary):
+    @classmethod
```
Comment (niosus) [inline plugin/tools.py]: Is there a reason why you need `cls` here? Seems redundant to me.

File: plugin/tools.py [hunk 2]
```diff
@@ -702,14 +702,61 @@ def get_clang_version_str(clang_binary):
...
+            at all.
+        """
         version_regex = re.compile("\d\.\d\.*\d*")
         match = version_regex.search(output_text)
         if match:
             version_str = match.group()
-            # TODO: we may need to revisit this and detect, say, word apple in
-            # the output of clang --version
-            if version_str >= "4.2" and platform.system() == "Darwin":
+            if version_str >= "4.2":
```
Comment (niosus) [inline plugin/tools.py]: We don't need this check now do we? We will only reach this when we know we are on osx and use apple clang, right?

File: plugin/tools.py [hunk 3]
```diff
@@ -731,6 +778,58 @@ def get_clang_version_str(clang_binary):
...
+        match = version_regex.search(output_text)
+        if match:
+            version_str = match.group()
+            return version_str
+        else:
+            raise RuntimeError(
+                " Couldn't find clang version in clang version output.")
+
+    @classmethod
+    def get_clang_version_str_windows(cls, output_text):
```
Comment (niosus) [inline plugin/tools.py]: Why so many functions? Searching for libraries was working uniformly across all platforms before. I see no reason in duplicating the code.

File: plugin/tools.py (hunk 1 above)
Comment (rwols) [inline plugin/tools.py]: I call into other class methods, seemed easier to reference this way

File: plugin/tools.py (hunk 2 above)
Comment (rwols) [inline plugin/tools.py]: I'm not sure, I'll check this out

File: plugin/tools.py (hunk 1 above)
Comment (niosus) [inline plugin/tools.py]: Ok, fair point. I agree.

File: plugin/tools.py (hunk 3 above)
Comment (rwols) [inline plugin/tools.py]: Yeah, windows and linux seem to behave the same for now. I'll change this to call into the linux variant.

**Comment by niosus** [comment]
And yeah, I don't know what's wrong with windows. The tests that fail are the ones responsible for checking that we remove smth like destructors or private methods from the output of completion function. It is strange that it fails as you don't seem to change anything that might have an influence on this.

File: plugin/tools.py (hunk 3 above)
Comment (niosus) [inline plugin/tools.py]: Thanks! :+1:

**Comment by niosus** [comment]
Also, @rwols please rebase onto `dev` branch. `master` is for urgent bug fixes. `dev` is for new developments. I don't always manage to have it that way, but at least I am trying.

File: plugin/tools.py [hunk 4]
```diff
@@ -702,34 +702,68 @@ def get_clang_version_str(clang_binary):
...
+
+    @classmethod
+    def _get_clang_version_str_osx(cls, output_text):
+        # There can be two flavours on osx: the apple flavour or the trunk/brew
+        # flavour.
+        if "Apple" in output_text:
+            return cls._get_apple_clang_version_str(output_text)
+        else:
+            # Same as on linux
+            return cls._get_clang_version_str_linux(output_text)
```
Comment (niosus) [inline plugin/tools.py]: I still don't really like having the name of the os in the function name. Basically, we have 2 ways to get the version: `clang` and `apple_clang`. We should have 2 functions: `_get_apple_clang_version_str` and `_get_clang_version_str`. Then all the boilerplate functions become unnecessary. These functions should just be called inline.

Comment (rwols) [inline plugin/tools.py]: Hmm, alright, I'll do this tomorrow.

**Comment by niosus** [comment]
Ok, so the issue with the windows build is that it fails to load libclang and falls back to using clang binary for completions. The key to finding that out is here:
```
======================================================================
FAIL: test_update (test_view_config.TestViewConfigManager)
Test that update is triggered.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\st\Data\Packages\EasyClangComplete\tests\test_view_config.py", line 162, in test_update
    self.assertEqual(view_config.completer.name, "lib")
AssertionError: 'bin' != 'lib'
- bin
+ lib
```

This means that you have changed something that touched the windows way of loading libclang. This is a very sensitive part of code and it took me hours if not days to get it right and to make sure we actually can use clang on all 3 systems. So it will take some time to get it right. Please try to run your code on windows or, if you don't have a windows machine set up with clang, look at your changes really carefully and see what could go wrong. Maybe revert back to the original code and try to only do small incremental changes, while making sure the code still builds in the CI. Good luck!

**Comment by rwols** [comment]
OK, I've removed the platform-splitting stuff. You might be interested in knowing that the constants at https://github.com/niosus/EasyClangComplete/pull/291/files#diff-d24472f3b095229b768dc8b58946b91dR769 are not actually defined anywhere (`PROGRESS_CHARS` and `PROGRESS_MSG`)

**Comment by niosus** [comment]
@rwols thanks for pointing that out. I removed them in favor of a dedicated class. I should also remove that function.

**Comment by niosus** [comment]
Now it looks good to me. Thanks for your contribution!

**Comment by rwols** [comment]
You're welcome :)

**Pull request merged**